pkg/csi/service/common/vsphereutil.go

```diff
@@ -605,10 +605,29 @@ func ExpandVolumeUtil(ctx context.Context, manager *Manager, volumeID string, ca
 		log.Infof("Successfully expanded volume for volumeid %q to new size %d Mb.", volumeID, capacityInMb)
 		return "", nil
 	} else {
-		expansionRequired, err := isExpansionRequired(ctx, volumeID, capacityInMb, manager, useAsyncQueryVolume)
+		var expansionRequired bool
+		vc, err := GetVCenter(ctx, manager)
 		if err != nil {
+			log.Errorf("failed to get vcenter. err=%v", err)
 			return csifault.CSIInternalFault, err
 		}
+		isvSphere8AndAbove, err := IsvSphere8AndAbove(ctx, vc.Client.ServiceContent.About)
+		if err != nil {
+			return "", logger.LogNewErrorf(log,
+				"Error while determining whether vSphere version is 8 and above %q, Error= %+v",
+				vc.Client.ServiceContent.About.ApiVersion, err)
+		}
+		// For vSphere 8 and above we avoid querying volume to check the size.
+		// This is handled internally in the FCD layer, CSI will just invoke
+		// ExtendDisk() all the time for vSphere version 8 and above.
+		if !isvSphere8AndAbove {
+			expansionRequired, err = isExpansionRequired(ctx, volumeID, capacityInMb, manager, useAsyncQueryVolume)
```

Comment:
We can club it in the existing check for idempotency on line#598
```
if isIdempotencyHandlingEnabled || isvSphere8AndAbove {
    // Call expand directly.
}

Reply:
+1

but one question, why  Idempotency feature assume we do not need to query volume ahead of extending?

Reply:
Resolved in PR https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1514.
@divyenpatel As far as I remember, improved idempotency feature stores the actual size of volume in a CR and compares against that value for an incoming API call. Therefore, I guess we avoid this check when it is enabled.